### PR TITLE
CAT-1397 Added Bricks to change Rotation Style

### PR DESCRIPTION
**Pull request by michael-fischl**
Finished from Lukas: refactored some stuff and adjusted tests as well as fixed a little bug; has Scratch behaviour

### CODE REVIEW

**Comment by robertpainsi** [comment]
Since it's a single commit PR, please amend the unformatted code changes.

**Comment by thmq** [comment]
@michael-fischl: could you please rebase this?

**Comment by michael-fischl** [comment]
with develop branch of upstream?

Am 03.03.16 um 14:22 schrieb Thomas Schranz:

File: catroid/res/layout/brick_set_rotation_style.xml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        android:id="@+id/brick_set_rotation_style_normal_layout"
+        style="@style/BrickContainer.Motion.Medium"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        app:horizontalSpacing="@dimen/brick_flow_layout_horizontal_spacing"
+        app:verticalSpacing="@dimen/brick_flow_layout_vertical_spacing" >
+
+        <TextView
+            android:id="@+id/brick_set_rotation_style_normal_label1"
```
Comment (84n4n4) [inline catroid/res/layout/brick_set_rotation_style.xml]: "id/brick_set_rotation_style_normal_text" would be more intuitive and way nicer than "id/brick_set_rotation_style_normal_label1"

**Comment by thmq** [comment]
ok to test

File: catroid/src/org/catrobat/catroid/content/Look.java [hunk 2]
```diff
@@ -303,12 +311,47 @@ public float getDirectionInUserInterfaceDimensionUnit() {
...
-		setRotation((-degrees + DEGREE_UI_OFFSET) % 360);
+		float oldRotation = getDirectionInUserInterfaceDimensionUnit();
+		rotation = (-degrees + DEGREE_UI_OFFSET) % 360;
+		realRotation = degrees;
+		float newRotation = getDirectionInUserInterfaceDimensionUnit();
+
+		switch (rotationMode) {
+			case ROTATION_STYLE_LEFT_RIGHT_ONLY:
+				setRotation(0f);
+				if ((oldRotation < 0 && newRotation >= 0) || (oldRotation >= 0 && newRotation < 0)) {
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/Look.java]: Can you use a boolean for this statement. It should improve the readability of the statement.

**Comment by bernadettespieler** [comment]
ok to test

**Comment by 84n4n4** [comment]
@michael-fischl: could you please rebase this?

**Comment by michael-fischl** [comment]
Already on it, seems tricky though since the ExtendedActions were 
removed by the Physics merge and I need to find out how all of this 
works now.

Am 28.04.16 um 13:14 schrieb Thomas Hirsch:

File: catroid/src/org/catrobat/catroid/content/actions/SetRotationStyleAction.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		try {
+			Integer newMode = (mode == null) ? Integer.valueOf(0) : mode.interpretInteger(sprite);
+
+			sprite.look.setRotationMode(newMode);
+			if (sprite.look.isFlipped()) {
+				sprite.look.getLookData().getTextureRegion().flip(true, false);
+				sprite.look.setFlipped(false);
+			}
+
+			if (!sprite.look.isFlipped() && sprite.look.getRealDirectionInUserInterfaceDimensionUnit() < 0 && newMode == 0) {
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/actions/SetRotationStyleAction.java]: Please use an boolean value in the if.. it's better readable

File: catroid/src/org/catrobat/catroid/content/actions/SetRotationStyleAction.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+				sprite.look.getLookData().getTextureRegion().flip(true, false);
+				sprite.look.setFlipped(true);
+			}
+			sprite.look.setDirectionInUserInterfaceDimensionUnit(sprite.look.getRealDirectionInUserInterfaceDimensionUnit());
+		} catch (InterpretationException interpretationException) {
+			Log.d(getClass().getSimpleName(), "Formula interpretation for this specific Brick failed.", interpretationException);
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/actions/SetRotationStyleAction.java]: Use Log.e for errormessages

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+import com.badlogic.gdx.scenes.scene2d.actions.SequenceAction;
+
+import org.catrobat.catroid.R;
+import org.catrobat.catroid.content.Sprite;
+import org.catrobat.catroid.content.actions.ExtendedActions;
+import org.catrobat.catroid.formulaeditor.Formula;
+
+import java.util.List;
+
+public class SetRotationStyleBrick extends BrickBaseType {
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Insert a newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+		if (!(checkbox.getVisibility() == View.VISIBLE)) {
+			spinner.setClickable(true);
+			spinner.setEnabled(true);
+		} else {
+			spinner.setClickable(false);
+			spinner.setEnabled(false);
+		}
+
+		final ArrayAdapter<String> spinnerAdapter = createSpinnerAdapter(context);
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+		SpinnerAdapterWrapper spinnerAdapterWrapper = new SpinnerAdapterWrapper(context, spinnerAdapter);
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+		spinner.setAdapter(spinnerAdapterWrapper);
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+				selection = position;
+			}
+
+			@Override
+			public void onNothingSelected(AdapterView<?> arg0) {
+			}
+		});
+
+		spinner.setSelection(selection, true);
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+	@Override
+	public Brick copyBrickForSprite(Sprite sprite) {
+		SetRotationStyleBrick copyBrick = (SetRotationStyleBrick) clone();
+		return copyBrick;
+	}
+
+	@Override
+	public View getViewWithAlpha(int alphaValue) {
+		if (view != null) {
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+		prototypeView = View.inflate(context, R.layout.brick_set_rotation_style, null);
+
+		spinner = (Spinner) prototypeView.findViewById(R.id.brick_set_rotation_style_spinner);
+		spinner.setFocusableInTouchMode(false);
+		spinner.setFocusable(false);
+		spinner.setEnabled(false);
+
+		SpinnerAdapter setLookSpinnerAdapter = createSpinnerAdapter(context);
+		spinner.setAdapter(setLookSpinnerAdapter);
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+	@Override
+	public Brick clone() {
+		return new SetRotationStyleBrick();
+	}
+
+	@Override
+	public List<SequenceAction> addActionToSequence(Sprite sprite, SequenceAction sequence) {
+		sequence.addAction(ExtendedActions.setRotationStyle(sprite, new Formula(selection)));
+		return null;
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Return null? Please return at least a empty list...

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+		protected Context context;
+		protected ArrayAdapter<String> spinnerAdapter;
+
+		private boolean isTouchInDropDownView;
+
+		public SpinnerAdapterWrapper(Context context, ArrayAdapter<String> spinnerAdapter) {
+			this.context = context;
+			this.spinnerAdapter = spinnerAdapter;
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+	public List<SequenceAction> addActionToSequence(Sprite sprite, SequenceAction sequence) {
+		sequence.addAction(ExtendedActions.setRotationStyle(sprite, new Formula(selection)));
+		return null;
+	}
+
+	private class SpinnerAdapterWrapper implements SpinnerAdapter {
+
+		protected Context context;
+		protected ArrayAdapter<String> spinnerAdapter;
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+		}
+
+		@Override
+		public boolean hasStableIds() {
+			return spinnerAdapter.hasStableIds();
+		}
+
+		@Override
+		public View getView(int paramInt, View paramView, ViewGroup paramViewGroup) {
+			if (isTouchInDropDownView) {
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Unnecessary if.. set isTouchInDropDownView to false is sufficient

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+
+		@Override
+		public View getDropDownView(int paramInt, View paramView, ViewGroup paramViewGroup) {
+			View dropDownView = spinnerAdapter.getDropDownView(paramInt, paramView, paramViewGroup);
+
+			dropDownView.setOnTouchListener(new View.OnTouchListener() {
+				@Override
+				public boolean onTouch(View paramView, MotionEvent paramMotionEvent) {
+					isTouchInDropDownView = true;
+					return false;
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Why return false?

File: catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+				}
+			});
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetRotationStyleBrick.java]: Remove newline

File: catroidTest/src/org/catrobat/catroid/test/content/actions/SetRotationStyleTest.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+import org.catrobat.catroid.formulaeditor.Formula;
+
+public class SetRotationStyleTest extends AndroidTestCase {
+
+	private Sprite sprite;
+
+	@Override
+	protected void setUp() throws Exception {
+		sprite = new Sprite("testSprite");
+		super.setUp();
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/test/content/actions/SetRotationStyleTest.java]: Call super first

File: catroidTest/src/org/catrobat/catroid/test/content/actions/SetRotationStyleTest.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	public void testNormalMode() {
+		ExtendedActions.setRotationStyle(sprite, new Formula(1));
+		ExtendedActions.pointInDirection(sprite, new Formula(Direction.RIGHT.getDegrees())).act(1.0f);
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/test/content/actions/SetRotationStyleTest.java]: Magic numbers.. Please put those numbers in a constant.

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/SetRotationStyleTest.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+	private Project project;
+
+	public SetRotationStyleTest() {
+		super(ScriptActivity.class);
+	}
+
+	@Override
+	public void setUp() throws Exception {
+		createProject();
+		super.setUp();
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/SetRotationStyleTest.java]: Call super first

**Comment by thmq** [comment]
please rebase and consider @thomasmauerhofer 's suggestions.

**Pull request closed without merging**